wrapper/src/main/java/software/amazon/jdbc/plugin/failover2/FailoverConnectionPlugin.java

```diff
@@ -0,0 +1,619 @@
+            this.pluginService.getCurrentHostSpec().getAliases(), HostAvailability.NOT_AVAILABLE);
+        try {
+          this.pickNewConnection();
+        } catch (final SQLException e) {
+          throw WrapperUtils.wrapExceptionIfNeeded(exceptionClass, e);
+        }
+        this.lastExceptionDealtWith = originalException;
+      }
+
+      if (originalException instanceof Error) {
+        throw (Error) originalException;
+      }
+      exceptionToThrow = originalException;
+    }
+
+    if (exceptionToThrow instanceof Error) {
+      throw (Error) exceptionToThrow;
+    }
+
+    throw WrapperUtils.wrapExceptionIfNeeded(exceptionClass, exceptionToThrow);
+  }
+
+  protected <E extends Exception> void dealWithIllegalStateException(
+      final IllegalStateException e,
+      final Class<E> exceptionClass) throws E {
+    dealWithOriginalException(e.getCause(), e, exceptionClass);
+  }
+
+  /**
+   * Initiates the failover procedure. This process tries to establish a new connection to an
+   * instance in the topology.
+   *
+   * @param failedHost The host with network errors.
+   * @throws SQLException if an error occurs
+   */
+  protected void failover(final HostSpec failedHost) throws SQLException {
+    this.pluginService.setAvailability(failedHost.asAliases(), HostAvailability.NOT_AVAILABLE);
+
+    if (this.failoverMode == FailoverMode.STRICT_WRITER) {
+      failoverWriter();
+    } else {
+      failoverReader();
+    }
+
+    if (isInTransaction || this.pluginService.isInTransaction()) {
+      if (this.pluginManagerService != null) {
+        this.pluginManagerService.setInTransaction(false);
+      }
+      // "Transaction resolution unknown. Please re-configure session state if required and try
+      // restarting transaction."
+      final String errorMessage = Messages.get("Failover.transactionResolutionUnknownError");
+      LOGGER.info(errorMessage);
+      throw new TransactionStateUnknownSQLException();
+    } else {
+      // "The active SQL connection has changed due to a connection failure. Please re-configure
+      // session state if required. "
+      LOGGER.severe(() -> Messages.get("Failover.connectionChangedError"));
+      throw new FailoverSuccessSQLException();
+    }
+  }
+
+  protected void failoverReader() throws SQLException {
+    TelemetryFactory telemetryFactory = this.pluginService.getTelemetryFactory();
+    TelemetryContext telemetryContext = telemetryFactory.openTelemetryContext(
+        TELEMETRY_READER_FAILOVER, TelemetryTraceLevel.NESTED);
+    this.failoverReaderTriggeredCounter.inc();
+
+    this.failoverStartTimeNano = System.nanoTime();
+
+    try {
+      LOGGER.fine(() -> Messages.get("Failover.startReaderFailover"));
+
+      // It's expected that this method synchronously returns when topology is stabilized,
+      // i.e. when cluster control plane has already chosen a new writer.
+      if (!this.pluginService.forceRefreshHostList(false, 0)) {
```

Comment:
why are we passing a 0 timeout instead of the failover timeout the user specified?

Reply:
Correct. We don't want to wait for a writer to be confirmed. We assume that a topology has other readers and we can try to connect to them.